### PR TITLE
debug-info: Make sure it is safe to remove the section

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

With this PR we check all the debug sections and only remove them if they do not have any bits. This clean up was introduced in https://github.com/parca-dev/parca-agent/pull/102 for another special case for Go binaries where we had  `.debug_gdb_scripts` section without any bits.

I hope the PR will be a solution for both.

Fixes https://github.com/parca-dev/parca-agent/issues/185

cc @derekparker

### CODE REVIEW

**Comment by brancz** [comment]
Happy with this being a follow-up, but could we have a test for this, maybe with an actual Rust binary?

**Comment by brancz** [comment]
On second thought I think I might prefer a real e2e test where we add containers to a kubernetes cluster of all kinds of languages.

**Comment by kakkoyun** [comment]
> On second thought I think I might prefer a real e2e test where we add containers to a kubernetes cluster of all kinds of languages.

I was going to suggest exactly this. We need a more flexible and extensible solution. Our support matrix will ever grow. I'm going to update the concerning issue with more context.

File: pkg/debuginfo/debuginfo.go [hunk 1]
```diff
@@ -243,18 +243,30 @@ func (di *Extractor) extract(ctx context.Context, buildID string, file string) (
 		level.Debug(di.logger).Log("msg", "failed to determine if binary is a Go binary", "path", file, "err", err)
 	}
 
+	sections, err := sectionsToRemove(file)
```
Comment (derekparker) [inline pkg/debuginfo/debuginfo.go:246]: nit: I would rename this to something more like `removeable` or `toRemove`.

File: pkg/debuginfo/debuginfo.go [hunk 2]
```diff
@@ -243,18 +243,30 @@ func (di *Extractor) extract(ctx context.Context, buildID string, file string) (
...
-		cmd, debugInfoFile = di.useStrip(ctx, tmpDir, file)
+		cmd, debugInfoFile, cleanup = di.useStrip(ctx, tmpDir, file, sections)
 	case isGo:
-		cmd, debugInfoFile = di.useObjcopy(ctx, tmpDir, file)
+		cmd, debugInfoFile, cleanup = di.useObjcopy(ctx, tmpDir, file, sections)
 	default:
-		cmd, debugInfoFile = di.useStrip(ctx, tmpDir, file)
+		cmd, debugInfoFile, cleanup = di.useStrip(ctx, tmpDir, file, sections)
 	}
+	defer func() {
```
Comment (derekparker) [inline pkg/debuginfo/debuginfo.go:264]: Why defer the cleanup?

Comment (derekparker) [reply pkg/debuginfo/debuginfo.go:264]: Also, why return a function at all? It seems that `useStrip` is the only method returning a cleanup function that does anything. Why not leave that defer in `useStrip` which cleans the interim file itself?

File: pkg/debuginfo/debuginfo.go [hunk 3]
```diff
@@ -347,49 +359,49 @@ func hasDebugInfo(path string) (bool, error) {
...
-		case strings.HasPrefix(s.Name, ".debug_"):
-			return s.Name[7:]
-		case strings.HasPrefix(s.Name, ".zdebug_"):
-			return s.Name[8:]
-		case strings.HasPrefix(s.Name, "__debug_"): // macos
-			return s.Name[8:]
-		default:
-			return ""
-		}
+var dwarfSuffix = func(s *elf.Section) string {
```
Comment (derekparker) [inline pkg/debuginfo/debuginfo.go:376]: nit: I'm being incredibly pedantic here but this should be `dwarfPrefix` as opposed to `dwarfSuffix`.

**Review by derekparker — COMMENTED** [review]
Just a few nits / questions.

File: pkg/debuginfo/debuginfo.go (hunk 3 above)
Comment (kakkoyun) [reply pkg/debuginfo/debuginfo.go:376]: This is actually, removing the detected prefix and returning the suffix. Thus the name. Maybe I should a comment to clarify more.

Comment (derekparker) [reply pkg/debuginfo/debuginfo.go:376]: Aah, ok I understand now. Sorry for the noise!

File: pkg/debuginfo/debuginfo.go (hunk 2 above)
Comment (derekparker) [reply pkg/debuginfo/debuginfo.go:264]: Ack, ok I see now, that makes sense.

However, is there any reason to not simply execute the command in the function we call and return the output directly? It seems like the caller (`di.extract` in this case) is only concerned with the output and doesn't need a handle to the `*exec.Command` pointer.

Comment (kakkoyun) [reply pkg/debuginfo/debuginfo.go:264]: I think this we can do :+1: Thanks for the suggestion.

**Comment by kakkoyun** [comment]
I have applied the suggestion @derekparker. PTAL.

**Review by derekparker — APPROVED** [review]
LGTM

**Pull request merged**